L3 shall rise again

sebageek opened the pull request.

sebageek committed:
Rename VXMAP to VXMAPs

We're normally talking about multiple vxmaps here

sebageek committed:
Remove duplicate switch config code from API

Originally the API was the only piece of code that generated whole
switch configs, but as we now need this code in other places this task
is now hanled by the FabricPlugin. As I somehow forgot to remove the old
code I'm doing that now.

sebageek committed:
Implement l3 functionality for external networks

With this commit we finally have l3 support. This includes the OpenStack
handling (creating the config) and the Arista config part.

For portbindings in external networks we now configure gateways on
routed vlan interfaces, and announce their networks/aggregates via BGP.
Relevant information to get this together is fetched by combining the
subnet's subnet pool, address scope and with that the address scope <->
VRF mapping from config.

On the Arista config side we now can configure routed Vlan interfaces,
prefix lists and BGP VRF aggregate sides.

swagner-de commented on networking_ccloud/common/config/config_driver.py:
```diff
@@ -505,6 +506,13 @@ def get_availability_zone(self, az_name):
                 return az
         return None
 
+    def get_vrf_name_for_address_scope(self, address_scope):
```
Maybe we could annotate with an lru_cache? Also, we should maybe add a validator on the address_scope_name uniqueness.

sebageek replied:
That sadly doesn't work with pydantic, so I opted for pregenerating the dicts. The problem was that pydantic no longer recognized the methods as methods.

swagner-de commented on networking_ccloud/db/db_plugin.py:
```diff
@@ -270,3 +296,110 @@ def remove_transit_from_network(self, context, network_id, az):
+            try:
+                if net_az_hint:
+                    net_az_hint = json.loads(net_az_hint)
+                    if len(net_az_hint) >= 1:
+                        az_hint = net_az_hint[0]
```
Do we make sure there will never be more than 1 az_hints? If so, why `>=` and not `==`?

sebageek replied:
Yeah, we have a `network_create_precommit()` hook that should prevent this. The `>=` is just there as a mean of "defensive programming", as I expect this to fail at some point because everything fails at some point and I'd not want the network to get regional then.

swagner-de commented on networking_ccloud/db/db_plugin.py:
```diff
@@ -270,3 +296,110 @@ def remove_transit_from_network(self, context, network_id, az):
+
+        snp_az = {}
+        for snp_id, tag in query.all():
+            if tag.startswith(cc_const.AZ_TAG_PREFIX):
```
```
query = query.filter(tag_models.Tag.tag.like(f'{cc_const.AZ_TAG_PREFIX}%'))

```

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -169,13 +219,17 @@ class VlanTranslation(pydantic.BaseModel):
     members: List[str] = None
 
+    # vlan interfaces only
+    vrf: str = None
```
Shouldn't this be on the `VlanIface`?

sebageek replied:
Hmyeah, we don't use that anymore.

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -355,7 +418,7 @@ def add_binding_host_to_config(self, hg_config, network_id, seg_vni, seg_vlan, t
                             iface.native_vlan = seg_vlan
 
-    def add_segments(self, net_segments, top_segments):
+    def add_segments(self, net_segments, top_segments, net_gateways):
```
I think this shoud take a `vrf` paramter as well and push it through to the `add_binding_host_to_config`.

sebageek replied:
vrf is now part of `net_gateways`

swagner-de reviewed (commented):
I think I am halfway through. I will continue Wednesday.

sebageek committed:
Fix l3 bgp lifecycle

VRF Networks (i.e. OpenStack subnet CIDRs) now are configured and
deconfigured together with the segment joining and leaving a physnet (in
most cases == switch). This makes sense, as only one subnet inside of a
VRF should be able to have a certain CIDR (if this changes, we have a
severely broken config and most likely other problems as well).
Aggregates on the other hand might have a different lifecycle, as
multiple networks in the same subnetpool might share some aggregates.
We now could do some DB queries for this and try to find out what the
situation is for a certain subnetpool and figure out what to add or
delete... But at the moment I want to get this out of the door,
therefore we choose the simple approach: Let it be. The switch syncloop
will take care of this every 15 minutes, so we have bogus aggregates for
a maximum time of 15 minutes and they should not break anything for us.
If we find out they do then we'll have to revisit this.

The original idea of the l3 lifecycle somehow assumed that one network
could be on the same physnet multiple times, which is not possible. I
probably got this mixed up with the aggregate lifecycle.

swagner-de commented on networking_ccloud/common/config/config_driver.py:
```diff
@@ -233,6 +233,12 @@ def __hash__(self) -> int:
                     self.vni, self.untagged, tuple(self.dhcp_relays)))
 
+    @pydantic.validator('networks')
+    def ensure_at_lest_one_address(cls, nets):
```
Objection, we defined Infra Networks as Layer 2 or Layer 3 objects. So an Infra Network without network and VRF bound could still be bound as flat L2 network.

sebageek replied:
You are right, I will bring that back.

swagner-de commented on networking_ccloud/common/config/config_driver.py:
```diff
@@ -248,10 +254,6 @@ def ensure_network(cls, net):
 
     @pydantic.root_validator
     def ensure_correct_value_combination(cls, values):
```
On acceptance of the Layer 2 Infra Network comment, these validations should come back.

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -341,6 +399,11 @@ def add_binding_host_to_config(self, hg_config, network_id, seg_vni, seg_vlan, t
 
+            # gateways
+            if gateways:
+                scu.add_vlan_iface(vlan=seg_vlan, vrf=gateways['vrf'], primary_ip=gateways['ips'][0],
```
We need at least a semi-stable way for the ordering of primary and secondary IPs. In this implmentation the order comes down all the way from the DB. With all the joins inbetween I don't think the order is stable. An unstable order leads to us swapping primary and secondary IP over and over, each time creating a hickup. We probably cannot prevent this completely but relying on the DB feels dangerous.

sebageek replied:
The method that fetches the gateways already sorts them by int value of their network ip, so we always have a stable ordering. Whenever the ordering changes we will reconfigure all ips and the ip with the lowest int value will be the primary ip.

sebageek committed:
Ignore l3 gateways if handle_all_l3_gateways is off

We now can turn of l3 gateway handling by default! This means the driver
can run in the same region as for example our ACI driver and let the ACI
driver be l3 gateway for everything.

sebageek committed:
Add API call to move gateways onto the fabric

If we don't handle all gateways we might want to move gateways in
certain AZs (or even one by one in an AZ) onto the fabric. This can now
be done via a network API call! We emit an event for
"cc-fabric-network-gateway" with BEFORE_UPDATE, so other drivers
currently handling the l3 gateway for that network can tear it down
before we set it up.

sebageek committed:
Ensure subnet-network AZ and subnetpool AZ match

When a new subnet is created in an external network we now check that
the subnet's network's AZ hint matches the AZ of the subnetpool (given
to the subnetpool by a tag). This check is only executed if the
subnetpool has an address scope, because a) we only need to enforce this
for networks that are also "present in a VRF", which is managed via
address scopes (else the subnet is really more of an internal flavor...)
and b) because it was a bit more convenient to implement for me (which
means that we can definitely change this implementation if we need to).

The config option subnet_subnetpool_az_check_enabled is added so we can
turn this feature off if it ever breaks something in prod and we quickly
need to disable it without doing a drive code change. I don't really
expect us needing it, but you never know.

sebageek committed:
Call network sync on subnet CRUD for external networks

If a subnet is created, deleted or gets its gateway changed we need to
sync the network to make sure our switch config reflects this change.

sebageek committed:
L3 SVI/BGP on BGWs/Transits + binding rework

We now setup l3 on BGWs and Tranists (i.e. all interconnects). This
means that we have to do the same SVI (vlan_ifaces) and BGP VRF setup
that we need to do everywhere else when an l3 port is bound.

This whole thing put me in front of the question "why do we have 13
places that all call add_binding_host_to_config() and do something only
slightly different"? As all of these places need similar l3 config, I
decided to rework most of the code that calls
add_binding_host_to_config() to now mostly call
add_segments_to_config(), which then pulls the necessary l3 config and
properly fills the SwitchConfigUpdateList with this. This was a medium
sized refactoring, which will hopefully make the whole code much more
maintainable in the future.

sebageek committed:
Add API call to get openstack network config

This API returns the OpenStack config for one specific network. It can
be used to see how networking-ccloud thinks a network should look like.

sebageek committed:
Handle PCs without member interfaces in EOS

Sometimes Port-Channels don't have members, which seems to result in an
non-existent members/member attribute in GNMI. To be prepared for that
we assume this is not set with some get()s in python.
